### PR TITLE
Include code quality checks in main workflow

### LINKED ISSUE #1409: Code formatting / linting is in the wrong place
ATM the GitHub actions workflow contains the following step:

```yaml
      - name: Lint with black and isort
        run: poetry run tox -e lint
```

...which is great, but it does nothing for the code that is in the repository, much of which doesn't conform to black's formatting rules and which brings up errors when you run flake8.

It would be better to have this step and the flake8 run as a pre-commit hook (e.g. using the [pre-commit tool](https://pre-commit.com)) and add a template to PRs to ensure that devs have run black before submitting their PR.

See [example PR template](https://github.com/kbase/narrative/blob/develop/.github/pull_request_template.md) and [a filled-out version](https://github.com/kbase/narrative/pull/3253).

### PR DESCRIPTION
**Pull request by pkalita-lbl**
Fixes #1409 

Currently we have a tox testenv called `lint` which runs `black` and `isort`. The main problem with that is those commands should be run _before_ committing. What should instead be running as part of the workflow is a tool like [Ruff](https://beta.ruff.rs/docs/) to flag formatting and code quality issues and discouraging merging code with problems into `main`.

These changes:
* Remove isort and add Ruff. Ruff essentially supersedes both `isort` and other tools we could have added like [Flake8](https://flake8.pycqa.org/en/latest/). 
* Add a new tox testenv called `format` to run `black` and `ruff --fix`. This will automatically format the codebase and fix any problems that Ruff identifies that are automatically fixable (for example removing unused imports). This can be run on-demand using `tox -e format` to reformat your code if your IDE is not set up to automatically format it correctly. Configuration for these tools has been added to `pyproject.toml`.
* Reformatted all files with `tox -e format`. This commit has been excluded from git blames using [`.git-blame-ignore-revs`](https://docs.github.com/en/repositories/working-with-files/using-files/viewing-a-file#ignore-commits-in-the-blame-view).
* Changed the existing `lint` tox testenv to run `black --check` and `ruff`; fixed existing errors. This will run in PR workflows to flag formatting and code quality problems.
* Removed the standalone codespell.yaml workflow file, added a `codespell` tox testenv, and added `tox -e codespell` to the main workflow.

**With these changes PRs will now fail if files are not formatted correctly or code quality issues are found.**

Future ideas: 

* This doesn't go quite as far as what was mentioned in #1409 which also talks about connecting these tools to pre-commit hooks, but I think I'd like to see if people find this to be too onerous before taking that additional step.
* Right now we're using Ruff's `E` and `F` [rule categories](https://beta.ruff.rs/docs/rules/) (their recommended starting point) along with `I` (import sorting). In the future we could start incrementally adding additional rules categories (e.g. [bugbear](https://beta.ruff.rs/docs/rules/#flake8-bugbear-b), [pep8-naming](https://beta.ruff.rs/docs/rules/#pep8-naming-n), etc)

### CODE REVIEW

**Comment by dalito** [comment]
Maybe use [ruff](https://beta.ruff.rs/docs/) instead of flake8 and isort? It is just great!

**Comment by pkalita-lbl** [comment]
@dalito I agree. I started looking at it right after opening this PR. Marking this as a draft in the meantime.

**Review by cmungall — APPROVED** [review]

**Review by hrshdhgd — APPROVED** [review]
Pretty cool. I'll have to look into adding `ruff` to the cookiecutter templates.

File: tox.ini [hunk 1]
```diff
@@ -1,40 +1,49 @@
...
+
+[testenv:codespell]
+description = Run spell checker.
+skip_install = true
+deps = 
+    codespell
+    tomli  # required for getting config from pyproject.toml
+commands = codespell {posargs}
+
+[testenv:format]
```
[the remark this answers is not in the document]
Comment (hrshdhgd) [reply tox.ini:31]: Seems like `testenv:format` actually fixes the code formatting and `lint` just does a check and throws an error if something seems unformatted, correct?

[the remark this answers is not in the document]
Comment (hrshdhgd) [reply tox.ini:31]: Haha...agree with (1)! Makes sense. Thanks!

**Comment by cthoyt** [comment]
@pkalita-lbl would be amazing if you could also apply black to all auto-generated files after generation

**Pull request merged**